### PR TITLE
feat: fetch node keys from subnet certificate

### PR DESCRIPTION
**Pull request by krpeacock**
# Description

Adds support for pulling subnet ID and node public keys out of a Certificate, and making them available for the `subnet` path in CanisterStatus.

Involves some refactoring of certificate decoding, so we introduce `lookupResultToBuffer` to constrain the `Cert.lookup` method to its existing return types. This means that a `HashTree` with type empty will be returned as `undefined`, preserving existing behavior. However, the `lookup_path` function will now return `HashTree`s.

This will support the upcoming certified queries feature

Fixes # (issue)

# How Has This Been Tested?

new tests, including fresh certificates from application and system subnets on local and mainnet, on `dfx 0.15.0`

# Checklist:

- [x] My changes follow the guidelines in [CONTRIBUTING.md](https://github.com/dfinity/agent-js/blob/main/CONTRIBUTING.md).
- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

File: packages/agent/src/agent/http/__certificates__/goldenCertificates.test.ts
Comment (dfx-json) [inline packages/agent/src/agent/http/__certificates__/goldenCertificates.test.ts:1]: are these like the golden tickets in willy wonka?

File: packages/agent/src/certificate.ts [hunk 1]
```diff
@@ -21,20 +21,35 @@ interface Cert {
...
+
+export { NodeId };
 
 export type HashTree =
-  | [NodeId.Empty]
-  | [NodeId.Fork, HashTree, HashTree]
-  | [NodeId.Labeled, ArrayBuffer, HashTree]
-  | [NodeId.Leaf, ArrayBuffer]
-  | [NodeId.Pruned, ArrayBuffer];
+  | [typeof NodeId.Empty]
```
Comment (dfx-json) [inline packages/agent/src/certificate.ts:37]: how did this not produce a type error before?

File: packages/agent/src/certificate.ts [hunk 2]
```diff
@@ -59,19 +74,38 @@ export function hashTreeToString(tree: HashTree): string {
...
+        return `sub(\n left:\n${indent(left)}\n---\n right:\n${indent(right)}\n)`;
+      } else {
+        throw new Error('Invalid tree structure for fork');
+      }
     }
     case NodeId.Labeled: {
-      const label = labelToString(tree[1]);
-      const sub = hashTreeToString(tree[2]);
-      return `label(\n label:\n${indent(label)}\n sub:\n${indent(sub)}\n)`;
+      if (tree[1] instanceof ArrayBuffer && tree[2] instanceof ArrayBuffer) {
```
Comment (dfx-json) [inline packages/agent/src/certificate.ts:86]: some links to the spec as a comment would be helpful here imo

**Review by dfx-json — APPROVED** [review]
Looks good. Nice work @krpeacock 👍

File: packages/agent/src/certificate.ts (hunk 1 above)
Comment (krpeacock) [reply packages/agent/src/certificate.ts:37]: The JS agent's decoding logic simply didn't cover this case. The `/subnet` path never decoded correctly. 

At some point, the entire certificate ought to be re-written, modeled after the rust library for comprehensiveness

File: packages/agent/src/certificate.ts (hunk 2 above)
Comment (krpeacock) [reply packages/agent/src/certificate.ts:86]: This entire utility function doesn't seem to work and it has no tests. I only modified it in order to make it type safe with the other lookup changes

File: packages/agent/src/certificate.ts [hunk 3]
```diff
@@ -172,7 +208,50 @@ export class Certificate {
...
+          'tdb26-jop6k-aogll-7ltgs-eruif-6kk7m-qpktf-gdiqx-mxtrf-vb5e6-eqe',
+        ).toUint8Array(),
+        certificate: new ArrayBuffer(0),
+      };
+    }
+    const nodeTree = lookup_path(['subnet', delegation?.subnet_id as ArrayBuffer, 'node'], tree);
+    const nodeForks = flatten_forks(nodeTree as HashTree) as HashTree[];
+    nodeForks.length;
+
+    this.#nodeKeys = nodeForks.map(fork => {
```
Comment (ShuoWangNSL) [inline packages/agent/src/certificate.ts:242]: Here `nodeKeys` is an array of public keys. 
When we verify the node signature, we know the node id and would like to find the corresponding public. For this purpose, should `nodeKeys`  be a map from node id to public key?

**Pull request merged**